mllib/src/test/scala/org/apache/spark/ml/feature/ImputerSuite.scala

```diff
@@ -0,0 +1,121 @@
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.spark.ml.feature
+
+import org.apache.spark.SparkFunSuite
+import org.apache.spark.ml.util.{DefaultReadWriteTest}
+import org.apache.spark.mllib.util.MLlibTestSparkContext
+import org.apache.spark.mllib.util.TestingUtils._
+import org.apache.spark.sql.Row
+
+class ImputerSuite extends SparkFunSuite with MLlibTestSparkContext with DefaultReadWriteTest {
+
+  test("Imputer for Double with default missing Value NaN") {
+    val df = sqlContext.createDataFrame( Seq(
+      (0, 1.0, 1.0, 1.0, 1.0),
+      (1, 1.0, 1.0, 1.0, 1.0),
+      (2, 3.0, 3.0, 3.0, 3.0),
+      (3, 4.0, 4.0, 4.0, 4.0),
+      (4, Double.NaN, 2.25, 1.0, 1.0 )
+    )).toDF("id", "value", "mean", "median", "mode")
+    Seq("mean", "median", "mode").foreach { strategy =>
+      val imputer = new Imputer().setInputCol("value").setOutputCol("out").setStrategy(strategy)
+      val model = imputer.fit(df)
+      model.transform(df).select(strategy, "out").collect()
+        .foreach { case Row(d1: Double, d2: Double) =>
+          assert(d1 ~== d2 absTol 1e-5, s"Imputer ut error: $d2 should be $d1")
+        }
+    }
+  }
+
+  test("Imputer for Double with missing Value -1.0") {
+    val df = sqlContext.createDataFrame( Seq(
+      (0, 1.0, 1.0, 1.0, 1.0),
+      (1, 1.0, 1.0, 1.0, 1.0),
+      (2, 3.0, 3.0, 3.0, 3.0),
+      (3, 4.0, 4.0, 4.0, 4.0),
+      (4, -1.0, 2.25, 1.0, 1.0 )
+    )).toDF("id", "value", "mean", "median", "mode")
+    Seq("mean", "median", "mode").foreach { strategy =>
+      val imputer = new Imputer().setInputCol("value").setOutputCol("out").setStrategy(strategy)
+        .setMissingValue(-1.0)
+      val model = imputer.fit(df)
+      model.transform(df).select(strategy, "out").collect()
+        .foreach { case Row(d1: Double, d2: Double) =>
+          assert(d1 ~== d2 absTol 1e-5, s"Imputer ut error: $d2 should be $d1")
+        }
+    }
+  }
+
+  test("Imputer for Int with missing Value -1") {
+    val df = sqlContext.createDataFrame( Seq(
+      (0, 1, 1, 1, 1),
+      (1, 3, 3, 3, 3),
+      (2, 10, 10, 10, 10),
+      (3, 10, 10, 10, 10),
+      (4, -1, 6, 3, 10)
+    )).toDF("id", "value", "mean", "median", "mode")
+
+    Seq("mean", "median", "mode").foreach { strategy =>
+      val imputer = new Imputer().setInputCol("value").setOutputCol("out").setStrategy(strategy)
+        .setMissingValue(-1)
+      val model = imputer.fit(df)
+      model.transform(df).select(strategy, "out").collect()
+        .foreach { case Row(d1: Int, d2: Int) =>
+          assert(d1 === d2, s"Imputer ut error: $d2 should be $d1")
+        }
+    }
+  }
+
```

Comment:
we should also have a test for a `non-NaN` missing value, but with `NaN` in the dataset, to check that "mean" and "median" behave as we expect.